### PR TITLE
Add client_server_test bazel rule

### PR DESCRIPTION
**Pull request by dajmaki**
This adds a generic mechanism for constructing a test that launches
a TCP server and runs a test-suite against it. The rule orchestrates
passing the port number from the server to the client via a temporary
file and takes care of killing the server when the client process exits.

Still to be done is figuring out a good way to pass additional arguments
to the server and client programs (where these arguments could be other
bazel rules).

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](./CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Add a line to the [release notes](./docs/source/support/release-notes.rst), if appropriate

### CODE REVIEW

File: bazel_tools/client_server_test/client_server_test.bzl [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+    ),
+    "server_args": attr.string_list(),
+  },
+)
+"""Create a client-server test.
+
+The rule takes a client and server executables and their
+arguments as parameters. The server port is passed via a
+temporary file, which is passed to the server executable via the
+"--port-file" parameter. The client application receives the port
```
Comment (aherrmann-da) [inline bazel_tools/client_server_test/client_server_test.bzl]: ~Out of curiousity, why is the port passed to the server in a file, and to the client as an argument?~

Got it, it's the server that writes its port into that file. The docstring is a bit ambiguous in this.

File: bazel_tools/client_server_test/client_server_test.bzl [hunk 2]
```diff
@@ -0,0 +1,87 @@
+# Copyright (c) 2019 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+# SPDX-License-Identifier: Apache-2.0
+
+load("@bazel_skylib//lib:shell.bzl", "shell")
```
Comment (aherrmann-da) [inline bazel_tools/client_server_test/client_server_test.bzl]: Is this still used? `shell.quote` seems to be inlined in `_expand_args`.

File: bazel_tools/client_server_test/tests/BUILD.bazel [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+  ],
+)
+
+client_server_test(
+  name = "test",
+  client = ":client",
+  client_args = [
+    "--foobar",
+    # Use the input file passed in via `data`.
+    #"bazel_tools/client_server_test/tests/client_input_file"
```
Comment (aherrmann-da) [inline bazel_tools/client_server_test/tests/BUILD.bazel]:
```suggestion
```

File: bazel_tools/client_server_test/tests/BUILD.bazel [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+    "$(location :client_input_file)"
```
Comment (aherrmann-da) [inline bazel_tools/client_server_test/tests/BUILD.bazel]: awesome!

File: bazel_tools/client_server_test/tests/BUILD.bazel [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+  ],
+  server = ":server",
```
Comment (timw-da) [inline bazel_tools/client_server_test/tests/BUILD.bazel]: Will server args be supported?

Comment (gleber-da) [reply bazel_tools/client_server_test/tests/BUILD.bazel]: They are: https://github.com/digital-asset/daml/pull/365/files#diff-6c30fe5cafc51117473d933c4db9e94aR24

File: bazel_tools/client_server_test/client_server_test.bzl (hunk 2 above)
Comment (dajmaki) [reply bazel_tools/client_server_test/client_server_test.bzl]: thx, removed.

File: bazel_tools/client_server_test/client_server_test.bzl [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+    ),
+    is_executable = True
+  )
+
+  runfiles = ctx.runfiles(files = [wrapper])
+  runfiles = runfiles.merge(ctx.attr._runner[DefaultInfo].default_runfiles)
+  runfiles = runfiles.merge(ctx.attr.client[DefaultInfo].default_runfiles)
+  runfiles = runfiles.merge(ctx.attr.server[DefaultInfo].default_runfiles)
+
+  return struct(
```
[the remark this answers is not in the document]
Comment (dajmaki) [reply bazel_tools/client_server_test/client_server_test.bzl]: thanks, fixed.

File: bazel_tools/client_server_test/client_server_test.bzl (hunk 1 above)
Comment (dajmaki) [reply bazel_tools/client_server_test/client_server_test.bzl]: fixed the comment.

**Pull request merged**